File: mobile/src/main/java/com/example/steven/stlbusarrivals/UI/Fragments/DetailsFragment.java
package com.example.steven.stlbusarrivals.UI.Fragments;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.example.steven.stlbusarrivals.Constants;
import com.example.steven.stlbusarrivals.DAO.DatabaseHelper;
import com.example.steven.stlbusarrivals.Model.Details;
import com.example.steven.stlbusarrivals.Model.TimeList;
import com.example.steven.stlbusarrivals.R;
import com.example.steven.stlbusarrivals.RequestSender;
import com.j256.ormlite.android.apptools.OpenHelperManager;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Observable;
import java.util.Observer;

/**
 * Created by Steven on 2016-01-28.
 */
public class DetailsFragment extends Fragment implements Observer{

    private DatabaseHelper databaseHelper = null;
    private String stopId, routeTag, stopName, routeName;
    TextView tv_routeName, tv_stopName, firstPrediction, secondPrediction;
    ArrayList<Details> detailsList;



    public DetailsFragment() {
        // Required empty public constructor
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        stopId = getArguments().getString("stopId");
        stopName = getArguments().getString("stopName");
        routeTag = getArguments().getString("routeTag");
        routeName = getArguments().getString("routeName");
    }

    @Override
    public void onResume(){
        super.onResume();
        String url = "http://webservices.nextbus.com/service/publicXMLFeed?command=predictions&a=stl&stopId=" + stopId + "&routeTag=" + routeTag;
        RequestSender request = new RequestSender(getActivity(),Constants.PREDICTION_XML, url);
        request.addObserver(this);
        request.sendRequest();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        ViewGroup v = (ViewGroup) inflater.inflate(R.layout.fragment_details, container, false);
        tv_routeName = (TextView)v.findViewById(R.id.tv_details_route_name);
        tv_stopName = (TextView)v.findViewById(R.id.tv_details_stop_name);
        firstPrediction = (TextView) v.findViewById(R.id.first_prediction);
        secondPrediction = (TextView) v.findViewById(R.id.second_prediction);
        tv_routeName.setText(routeName);
        tv_stopName.setText(stopName);

        detailsList = getHelper().getAllOrderedDetails();

        final FloatingActionButton fab = (FloatingActionButton) v.findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Details details = new Details(getActivity());
                details.setTag(routeTag);
                details.setStopId(stopId);
                details.setRouteName(routeName);
                details.setStopName(stopName);
                getHelper().getDetailsDao().create(details);
                Toast.makeText(getActivity(), " Stop added to favorites!", Toast.LENGTH_SHORT).show();
                fab.setVisibility(View.GONE);
            }
        });
        for(int i=0; i<detailsList.size();i++){
            if(detailsList.get(i).getStopId().equals(stopId)){
                fab.setVisibility(View.GONE);
            }
        }
        return v;
    }

    @Override
     public void update(Observable observable, Object o) {
        Log.d("detailsfrag update", "entered");
        if(o instanceof TimeList){
            TimeList timeList = (TimeList) o;
            Calendar c = Calendar.getInstance();
            int currentHour = c.get(Calendar.HOUR_OF_DAY);
            int currentMinutes = c.get(Calendar.MINUTE);
            if(timeList.size() !=0) {
                int predictedMinutes = currentMinutes + Integer.valueOf(timeList.get(0).getTime());
                while(predictedMinutes >= 60){
                    currentHour++;
                    predictedMinutes = predictedMinutes - 60;
                }
                if(predictedMinutes < 10){
                    firstPrediction.setText(currentHour + ":0"+ predictedMinutes + "   Next bus is in " + timeList.get(0).getTime() + " minutes");
                }else {
                    firstPrediction.setText(currentHour + ":" + predictedMinutes + "   Next bus is in " + timeList.get(0).getTime() + " minutes");
                }
                if (timeList.size() > 1) {
                    int nextHour = c.get(Calendar.HOUR_OF_DAY);
                    int nextMinutes = c.get(Calendar.MINUTE);
                    int nextPredictedMinutes = nextMinutes + Integer.valueOf(timeList.get(1).getTime());
                    while(nextPredictedMinutes >= 60){
                        nextHour++;
                        nextPredictedMinutes = nextPredictedMinutes - 60;
                    }
                    if(nextPredictedMinutes <10){
                        secondPrediction.setText(nextHour + ":0"+ nextPredictedMinutes + "   Other bus is in " + timeList.get(1).getTime() + " minutes");
                    }else {
                        secondPrediction.setText(nextHour + ":" + nextPredictedMinutes + "   Other bus is in " + timeList.get(1).getTime() + " minutes");
                    }
                }
            }
        }
    }

    //Needed so that databaseHelper can be initialised
    private DatabaseHelper getHelper(){
        if(databaseHelper == null){
            databaseHelper = OpenHelperManager.getHelper(getActivity(),DatabaseHelper.class);
        }
        return databaseHelper;
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
        if(databaseHelper != null){
            OpenHelperManager.releaseHelper();
            databaseHelper = null;
        }
    }
}

File: mobile/src/main/java/com/tilly/steven/stlbusarrivals/ui/fragments/DetailsFragment.kt
package com.tilly.steven.stlbusarrivals.ui.fragments

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.tilly.steven.stlbusarrivals.*
import com.tilly.steven.stlbusarrivals.Utils.toast
import com.tilly.steven.stlbusarrivals.model.Details
import com.tilly.steven.stlbusarrivals.model.TimeList
import java.util.*


/**
 * Created by Steven on 2016-01-28.
 */
class DetailsFragment : Fragment(), Observer {

    private val xmlparser = XmlParser()
    private var stopId: String = ""
    private var routeTag: String = ""
    private var stopName: String = ""
    private var routeName: String = ""
    lateinit var tv_routeName: TextView
    lateinit var tv_stopName: TextView
    lateinit var firstPrediction: TextView
    lateinit var secondPrediction: TextView
    lateinit var tvMessage: TextView
    lateinit var detailsList: List<Details>
    private lateinit var mHandler: Handler
    lateinit var fab: FloatingActionButton

    private var mStatusChecker: Runnable = object : Runnable {
        override fun run() {
            sendRequest()
            val mInterval = 60000
            mHandler.postDelayed(this, mInterval.toLong())
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        xmlparser.addObserver(this)
        stopId = arguments?.getString("stopId") ?: ""
        stopName = arguments?.getString("stopName") ?: ""
        routeTag = arguments?.getString("routeTag") ?: ""
        routeName = arguments?.getString("routeName") ?: ""
        mHandler = Handler()
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val v = inflater.inflate(R.layout.fragment_details, container, false) as ViewGroup
        tv_routeName = v.findViewById(R.id.tv_details_route_name)
        tv_stopName = v.findViewById(R.id.tv_details_stop_name)
        firstPrediction = v.findViewById(R.id.first_prediction)
        secondPrediction = v.findViewById(R.id.second_prediction)
        tvMessage = v.findViewById(R.id.tv_messages)
        tv_routeName.text = routeName
        tv_stopName.text = stopName

        fab = v.findViewById(R.id.fab)
        fab.setOnClickListener {
            val details = Details()
            details.tag = routeTag
            details.stopId = stopId
            details.routeName = routeName
            details.stopName = stopName
            launchUI {
                asyncAwait { DetailsDatabase.getInstance().detailsDao().insertOrUpdate(details) }
                toast(message = getString(R.string.added_favorites))
                fab.visibility = View.GONE
            }
        }
        DetailsDatabase.getInstance().detailsDao().loadDetails().observe(this, androidx.lifecycle.Observer {
            detailsList = it
            for (i in detailsList.indices) {
                if (detailsList[i].stopId == stopId) {
                    fab.visibility = View.GONE
                }
            }
        })
        sendMessageRequest()
        return v
    }

    override fun onResume() {
        super.onResume()
        startRepeatingTask()
    }

    private fun startRepeatingTask() {
        mStatusChecker.run()
    }

    override fun onStop() {
        super.onStop()
        mHandler.removeCallbacks(mStatusChecker)
    }

    private fun sendRequest() {
        activity?.let {
            val queue = VolleySingleton.getInstance(it).requestQueue
            val url = "http://webservices.nextbus.com/service/publicXMLFeed?command=predictions&a=stl&stopId=$stopId&routeTag=$routeTag"
            val request = StringRequest(url, Response.Listener { response ->
                // we got the response, now our job is to handle it
                //parseXmlResponse(response);
                try {
                    xmlparser.readPrediction(response)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }, Response.ErrorListener { error ->
                error.printStackTrace()
                sendRequest()
            })
            queue.add(request)
        }
    }

    private fun sendMessageRequest() {
        activity?.let {
            val queue = VolleySingleton.getInstance(it).requestQueue
            val url = "http://webservices.nextbus.com/service/publicXMLFeed?command=messages&a=stl&r=$routeTag"
            val request = StringRequest(url, Response.Listener { response ->
                // we got the response, now our job is to handle it
                //parseXmlResponse(response);
                try {
                    xmlparser.readMessages(response)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }, Response.ErrorListener { error ->
                error.printStackTrace()
                sendMessageRequest()
            })
            queue.add(request)
        }
    }

    override fun update(observable: Observable, o: Any) {
        if (activity != null) {
            Log.d("detailsfrag update", "entered")
            if (o is TimeList) {
                val c = Calendar.getInstance()
                var currentHour = c.get(Calendar.HOUR_OF_DAY)
                val currentMinutes = c.get(Calendar.MINUTE)
                if (o.size != 0) {
                    var predictedMinutes = currentMinutes + Integer.valueOf(o[0].time)
                    while (predictedMinutes >= 60) {
                        currentHour++
                        predictedMinutes -= 60
                    }
                    if (predictedMinutes < 10) {
                        firstPrediction.text = "$currentHour:0$predictedMinutes   ${getString(R.string.next_bus, o[0].time)}"
                    } else {
                        firstPrediction.text = "$currentHour:$predictedMinutes   ${getString(R.string.next_bus, o[0].time)}"
                    }
                    if (o.size > 1) {
                        var nextHour = c.get(Calendar.HOUR_OF_DAY)
                        val nextMinutes = c.get(Calendar.MINUTE)
                        var nextPredictedMinutes = nextMinutes + Integer.valueOf(o[1].time)
                        while (nextPredictedMinutes >= 60) {
                            nextHour++
                            nextPredictedMinutes -= 60
                        }
                        if (nextPredictedMinutes < 10) {
                            secondPrediction.text = "$nextHour:0$nextPredictedMinutes   ${getString(R.string.other_bus, o[1].time)}"
                        } else {
                            secondPrediction.text = "$nextHour:$nextPredictedMinutes   ${getString(R.string.other_bus, o[1].time)}"
                        }
                    } else
                        secondPrediction.visibility = View.GONE
                }
            } else if (o is String) {
                tvMessage.text = o
            }
        }
    }
}
